Add first_line_match for content based recognition

scop opened the pull request:
Similar considerations re whitespace and major versions only as in #1

scop committed:
Add first_line_match for content based recognition

keith-hall commented on http-request-response.sublime-syntax:
```diff
@@ -5,6 +5,7 @@ scope: text.http-request-response
 name: HTTP Request and Response
 file_extensions:
   - http
+first_line_match: '^(?:(?:GET|HEAD|POST|PUT|DELETE|CONNECT|OPTIONS|TRACE|PATCH) .+ HTTP/\d(?:\.\d)?$|HTTP/\d(?:\.\d)? \d{3}(?: .*?)?$)'
```
I think this would be more readable if we use extended regex mode `(?x)` combined with multi-line yaml, like was done in this PR:
https://github.com/sublimehq/Packages/pull/2876/files#diff-1a0bf514090420354f8e65d3e57b6906366c93c0d6b9734122f65d8934c98b80R7-R10

Also, I think we don't need the `(?: .*?)?$` part, as the `first_line_match` doesn't have to match the whole line, and it could perhaps be more efficient without it. We could make the space after the status code mandatory, and it would have the same behavior if I understand correctly while still working for the `curl` debug output mentioned in #1.

scop replied:
Good idea, done in 924ca0897052fabaf9bc1a7bc7be256bbe99853c. Requiring the space after the status code is fine and enough, as it's required per the HTTP 1.0 and 1.1 spec, too (while the reason phrase following it can be empty).

scop replied:
BTW, the list of possible HTTP methods is not limited to those defined in the HTTP spec, see https://www.iana.org/assignments/http-methods and https://tools.ietf.org/html/rfc7231#section-4.1 (`token`, standardized ones uppercase ASCII by convention). So instead of enumerating methods, should we just accept something like `[A-Z]+(?:[A-Z-]*[A-Z]+)?` (not quite the full range of uppercase `token`, but you get the gist).

scop replied:
(If you like ^that idea, I suggest doing it in a separate followup commit after this PR, as it applies to the first match in `main`, too.)

keith-hall replied:
Sounds like a plan. It might be nice to scope the common http verbs differently, however compared to custom ones, in case people want to color them differently or something.

keith-hall reviewed (changes requested):
Thanks for this, definitely very useful to have - not sure how I missed it in the first place, sorry about that! I have some minor comments about how to make the regex easier to understand and perhaps quicker for regex engines to match against.

keith-hall commented:
Note: I also logged https://github.com/sublimehq/sublime_text/issues/4713 because I saw an opportunity for the `first_line_match` to become even easier to maintain :)
Once ST implements it, I can update `syntect` also :)

scop committed:
Simplify first_line_match

Make use of YAML multiline strings and ?x regular expression syntax,
require simply a space after status code for responses.

https://tools.ietf.org/html/rfc7230#section-3.1.2

keith-hall commented:
Thanks!
